File: csharp/1575.count-all-possible-routes.cs
/*
 * @lc app=leetcode id=1575 lang=csharp
 *
 * [1575] Count All Possible Routes
 */

using System;
using System.Linq;

// @lc code=start
public class Solution {

    private static long CountRoutes(int[] locations, int n, int t, int i, int f, long?[,] memo) {
        if (Math.Abs(locations[t] - locations[i]) > f) {
            return 0;
        }
        if (memo[i, f].HasValue) {
            return memo[i, f].Value;
        }
        long ret = t == i ? 1 : 0;
        for (int j = 0; j < n; ++j) {
            if (i == j) {
                continue;
            }
            int d = Math.Abs(locations[j] - locations[i]);
            if (d <= f) {
                ret = (ret + CountRoutes(locations, n, t, j, f - d, memo)) % 1000000007;
            }
        }
        memo[i, f] = ret;
        return ret;
    }

    public int CountRoutes(int[] locations, int start, int finish, int fuel) {
        int n = locations.Length;
        return (int)CountRoutes(locations, n, finish, start, fuel, new long?[n, fuel + 1]);
    }
}
// @lc code=end

File: java/1575.count-all-possible-routes.java
/*
 * @lc app=leetcode id=1575 lang=java
 *
 * [1575] Count All Possible Routes
 */

// @lc code=start
class Solution {
    public int countRoutes(int[] locations, int start, int finish, int fuel) {
        int n = locations.length;
        return countRoutes(locations, finish, start, fuel, new Integer[n][fuel + 1]);
    }

    private static int countRoutes(int[] locations, int finish, int curr, int fuel, Integer[][] memo) {
        if (Math.abs(locations[curr] - locations[finish]) > fuel) {
            return 0;
        }
        Integer m = memo[curr][fuel];
        if (m != null) {
            return m;
        }
        int n = locations.length;
        long ret = curr == finish ? 1 : 0;
        for (int next = 0; next < n; ++next) {
            if (next == curr) {
                continue;
            }
            int d = Math.abs(locations[curr] - locations[next]);
            if (d > fuel) {
                continue;
            }
            ret = (ret + (long)countRoutes(locations, finish, next, fuel - d, memo)) % 1000000007;
        }
        memo[curr][fuel] = (int)ret;
        return (int)ret;
    }
}
// @lc code=end

